waku/v2/protocol/waku_store/waku_store_types.nim

```diff
@@ -113,30 +117,272 @@ type
+  var
+    w = SortedSetWalkRef[Index,IndexedWakuMessage].init(storeQueue.items)
+    res = w.last
+  w.destroy
+
+  if res.isOk:
+    return ok(res.value.data)
+  else:
+    return err("Not found")
+
+## --- Queue API ---
+
+proc new*(T: type StoreQueueRef, capacity: int): T =
+  var items = SortedSet[Index, IndexedWakuMessage].init()
+
+  return StoreQueueRef(items: items, capacity: capacity)
+
+proc add*(storeQueue: StoreQueueRef, msg: IndexedWakuMessage) =
   ## Add a message to the queue.
   ## If we're at capacity, we will be removing,
-  ## the oldest item
-
+  ## the oldest (first) item
+  
+  # TODO the below delete block can be removed if we convert to circular buffer
   if storeQueue.items.len >= storeQueue.capacity:
-    storeQueue.items.delete 0, 0  # Remove first item in queue
+    var
+      w = SortedSetWalkRef[Index, IndexedWakuMessage].init(storeQueue.items)
+      toDelete = w.first
+    discard storeQueue.items.delete(toDelete.value.key)
+    w.destroy # better to destroy walker after a delete operation
   
-  storeQueue.items.add(msg)
+  let res = storeQueue.items.insert(msg.index)
+  if res.isErr:
+    # return error result and log in metrics
+    echo "ERROR"
+  else:
+    res.value.data = msg
 
-proc len*(storeQueue: StoreQueue): int {.noSideEffect.} =
-  storeQueue.items.len
+proc getPage*(storeQueue: StoreQueueRef,
+              pred: QueryFilterMatcher,
+              pagingInfo: PagingInfo):
+             (seq[WakuMessage], PagingInfo, HistoryResponseError) {.gcsafe.} =
+  ## Get a single page of history matching the predicate and
+  ## adhering to the pagingInfo parameters
+  
+  let
+    cursorOpt = if pagingInfo.cursor == Index(): none(Index) ## TODO: pagingInfo.cursor should be an Option. We shouldn't rely on empty initialisation to determine if set or not!
+                else: some(pagingInfo.cursor)
+    maxPageSize = if pagingInfo.pageSize == 0 or pagingInfo.pageSize > MaxPageSize: MaxPageSize # Used default MaxPageSize for invalid pagingInfos
+                  else: pagingInfo.pageSize
+  
+  case pagingInfo.direction
+    of FORWARD:
+      return storeQueue.fwdPage(pred, maxPageSize, cursorOpt)
+    of BACKWARD:
+      return storeQueue.bwdPage(pred, maxPageSize, cursorOpt)
 
-proc allItems*(storeQueue: StoreQueue): seq[IndexedWakuMessage] =
-  storeQueue.items
+proc getPage*(storeQueue: StoreQueueRef,
+              pagingInfo: PagingInfo):
+             (seq[WakuMessage], PagingInfo, HistoryResponseError) {.gcsafe.} =
+  ## Get a single page of history without filtering.
+  ## Adhere to the pagingInfo parameters
+  
+  proc predicate(i: IndexedWakuMessage): bool = true # no filtering
+
+  return getPage(storeQueue, predicate, pagingInfo)
 
-template filterIt*(storeQueue: StoreQueue, pred: untyped): untyped =
-  storeQueue.items.filterIt(pred)
+proc len*(storeQueue: StoreQueueRef): int {.noSideEffect.} =
+  storeQueue.items.len
```

Comment:
I have no strong opinions here, but somewhat prefer the readability of expression-based returns for simple `proc`s such as these. These are quite canonical (see e.g. [the style guide on this topic](https://status-im.github.io/nim-style-guide/language.result.html?highlight=expression-based#result-return-languageresult) - we only try to avoid `result` returns). In this case, expression-based return is also consistent with [the submodule](https://github.com/status-im/nim-stew/blob/master/stew/sorted_set.nim#L185-L208).

Reply:
I have no strong opinion either, just though we started using explicit `return`s since the refactoring week.